### PR TITLE
Add tests for transaction dry run endpoint - Closes #1283

### LINKED ISSUE #1283: Add tests for transaction dry run endpoint
### Description

Lisk Service adds support to let end users dry run transactions with issues https://github.com/LiskHQ/lisk-service/issues/1280. The implementation needs to be tested on CI to ensure code quality.

### Acceptance Criteria

- Unit tests are implemented
- Integration tests are implemented

#### Endpoints
**HTTP**:  `POST /transactions/dryrun`
**RPC**:    `post.transactions.dryrun`

### PR DESCRIPTION
**Pull request by priojeetpriyom**
### What was the problem?
This PR resolves #1283

### How was it solved?
- [x] Unit tests are implemented - not necessary
- [x] Integration tests are implemented

### How was it tested?
- [x] Local

### CODE REVIEW

File: tests/schemas/api_v3/dryrunTransaction.schema.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import Joi from 'joi';
+
+const regex = require('./regex');
+
+const eventItem = {
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/dryrunTransaction.schema.js:20]: Can we reuse the `event` schema here? Also, `data` ideally should be an object.

File: tests/schemas/api_v3/dryrunTransaction.schema.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	data: Joi.string().pattern(regex.HEX).required(),
+	index: Joi.number().integer().positive().allow(0)
+		.required(),
+	module: Joi.string().min(1).required(),
+	name: Joi.string().min(1).required(),
+	topics: Joi.array().items(Joi.string().pattern(regex.HEX).required()).required(),
+	height: Joi.number().required(),
+};
+
+const dryrunTransactionSchema = {
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/dryrunTransaction.schema.js:30]: Let us rename it to `dryrunTransactionResponseSchema`.
```suggestion
const dryrunTransactionResponseSchema = {
```

File: tests/schemas/api_v3/dryrunTransaction.schema.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	success: Joi.boolean().valid(true).required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/dryrunTransaction.schema.js:31]: The success property can bear both `true` and `false`. So, please remove the `valid(true)` restriction here.
```suggestion
	success: Joi.boolean().required(),
```

File: tests/schemas/api_v3/dryrunTransaction.schema.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	events: Joi.array().items(Joi.object(eventItem).required()).required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/dryrunTransaction.schema.js:32]: Please set `const event = Joi.object({<schemaDetails>});` above;
```suggestion
	events: Joi.array().items(event).required(),
```

File: tests/integration/api_v3/rpc/dryRunTransactions.test.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+			{
+				transaction: VALID_TRANSACTION,
+			},
+		);
+		expect(response).toMap(jsonRpcEnvelopeSchema);
+
+		const { result } = response;
+		expect(result).toMap(goodRequestSchema);
+		expect(result.data).toBeInstanceOf(Object);
+		expect(result.data).toMap(dryrunTransactionSchema);
```
Comment (sameersubudhi) [inline tests/integration/api_v3/rpc/dryRunTransactions.test.js:50]: Please apply the same in the HTTP tests.
```suggestion
		expect(result.data).toMap(dryrunTransactionSchema);
		if (result.data.success === true) {
			expect(result.data.events.length).toBeGreaterThan(0);
		} else {
			expect(result.data.events.length).toBe(0);
		}
```

File: tests/integration/api_v3/rpc/dryRunTransactions.test.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	dryrunTransactionSchema,
+	goodRequestSchema,
+	metaSchema,
+} = require('../../../schemas/api_v3/dryrunTransaction.schema');
+
+const wsRpcUrl = `${config.SERVICE_ENDPOINT}/rpc-v3`;
+const postDryrunTransaction = async params => request(wsRpcUrl, 'post.transactions.dryrun', params);
+
+describe('Method post.transactions.dryrun', () => {
+	it('Post dryrun transaction succesfully', async () => {
```
Comment (sameersubudhi) [inline tests/integration/api_v3/rpc/dryRunTransactions.test.js:39]: Could also add another test case where we dry run an existing transaction? The expected `response.data` should look something like:
```
{
	 success: false,
	 events: [],
}
```

File: tests/integration/api_v3/http/dryRunTransactions.test.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	goodRequestSchema,
+	dryrunTransactionSchema,
+	metaSchema,
+} = require('../../../schemas/api_v3/dryrunTransaction.schema');
+
+const baseUrl = config.SERVICE_ENDPOINT;
+const baseUrlV3 = `${baseUrl}/api/v3`;
+const endpoint = `${baseUrlV3}/transactions/dryrun`;
+
+describe('Post dryrun transactions API', () => {
```
Comment (sameersubudhi) [inline tests/integration/api_v3/http/dryRunTransactions.test.js:35]: Please apply the corresponding changes requested in `tests/integration/api_v3/rpc/dryRunTransactions.test.js`.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: tests/schemas/api_v3/dryrunTransaction.schema.js (hunk 1 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/dryrunTransaction.schema.js:20]: The structure in these two differs.

File: tests/schemas/api_v3/dryrunTransaction.schema.js (hunk 3 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/dryrunTransaction.schema.js:31]: For failure the response structure is different. We only return this structure for success=true case. Should I still remove it?

File: tests/schemas/api_v3/dryrunTransaction.schema.js (hunk 2 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/dryrunTransaction.schema.js]: applied.

File: tests/schemas/api_v3/dryrunTransaction.schema.js (hunk 4 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/dryrunTransaction.schema.js]: applied.

File: tests/integration/api_v3/rpc/dryRunTransactions.test.js (hunk 5 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/rpc/dryRunTransactions.test.js]: applied.

File: tests/schemas/api_v3/regex.js [hunk 8]
```diff
@@ -31,6 +31,8 @@ const CHAIN_ID = /^\b[a-fA-F0-9]{8}\b$/;
 const TOKEN_ID = /^\b[a-fA-F0-9]{16}\b$/;
 const DURATION = /^\d{4}-\d{2}(?:-\d{2})?$/;
 const DIGITS = /^\d+$/;
+const HEX = /^\b[0-9a-fA-F]+\b$/;
+const TOPIC = /^\b(?:[0-9a-fA-F]+|lsk[a-hjkm-z2-9]{38})\b$/;
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/regex.js:35]: Could you please also add the same regex in gateway for both the `/events` and `/transactions/dryrun` endpoints?

File: tests/schemas/api_v3/event.schema.js [hunk 9]
```diff
@@ -29,12 +29,12 @@ const block = {
...
 	id: Joi.string().required(),
-	moduleID: Joi.string().required(),
-	moduleName: Joi.string().required(),
-	typeID: Joi.string().required(),
-	data: Joi.string().required(),
-	topics: Joi.array().required(),
-	block: Joi.object(block).optional(),
+	module: Joi.string().pattern(regex.MODULE).required(),
+	name: Joi.string().required(),
+	index: Joi.number().positive().required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/event.schema.js:34]: Index value starts from `0`. `positive()` enforces values to be in the range `[1, ∞)`. Please adjust accordingly. Kindly test the following suggestion.
```suggestion
	index: Joi.number().integer().min(0).required(),
```

File: tests/schemas/api_v3/event.schema.js [hunk 10]
```diff
@@ -29,12 +29,12 @@ const block = {
...
 const eventSchema = {
 	id: Joi.string().required(),
-	moduleID: Joi.string().required(),
-	moduleName: Joi.string().required(),
-	typeID: Joi.string().required(),
-	data: Joi.string().required(),
-	topics: Joi.array().required(),
-	block: Joi.object(block).optional(),
+	module: Joi.string().pattern(regex.MODULE).required(),
+	name: Joi.string().required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/event.schema.js:33]: Can we please add the pattern check for `name` as well?

File: tests/schemas/api_v3/dryrunTransaction.schema.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+const regex = require('./regex');
+
+const event = {
+	data: Joi.string().pattern(regex.HEX).required(),
+	index: Joi.number().integer().positive().allow(0)
+		.required(),
+	module: Joi.string().min(1).required(),
+	name: Joi.string().min(1).required(),
+	topics: Joi.array().items(Joi.string().pattern(regex.HEX).required()).required(),
+	height: Joi.number().required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/dryrunTransaction.schema.js:27]:
```suggestion
	data: Joi.string().object().required(),
	index: Joi.number().integer().min(0).required(),
	module: Joi.string().pattern(regex.MODULE).required(),
	name: Joi.string().pattern(regex.NAME).required(),
	topics: Joi.array().items(Joi.string().pattern(regex.TOPIC)).required(),
	height: Joi.number().positive().required(),
```

File: tests/schemas/api_v3/dryrunTransaction.schema.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+};
+
+const dryrunTransactionResponseSchema = {
+	success: Joi.boolean().required(),
+	events: Joi.array().items(Joi.object(event).optional()).required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/dryrunTransaction.schema.js:32]:
```suggestion
	events: Joi.array().items(Joi.object(event)).required(),
```

File: tests/integration/api_v3/rpc/dryRunTransactions.test.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+		expect(firstResponse).toMap(jsonRpcEnvelopeSchema);
+
+		const { result: firstResult } = firstResponse;
+		expect(firstResult).toMap(goodRequestSchema);
+		expect(firstResult.data).toBeInstanceOf(Object);
+		expect(firstResult.data).toMap(dryrunTransactionResponseSchema);
+		expect(firstResult.meta).toMap(metaSchema);
+		expect(firstResult.data.events.length).toBeGreaterThan(0);
+
+		// Send transaction and wait to be minted
```
Comment (sameersubudhi) [inline tests/integration/api_v3/rpc/dryRunTransactions.test.js:72]:
```suggestion
		// Send transaction and wait for it to be included in the next block
```

File: tests/integration/api_v3/rpc/dryRunTransactions.test.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+		const { result } = response;
+		expect(result).toMap(goodRequestSchema);
+		expect(result.data).toBeInstanceOf(Object);
+		expect(result.data).toMap(dryrunTransactionResponseSchema);
+		expect(result.meta).toMap(metaSchema);
+		expect(result.data.events.length).toBeGreaterThan(0);
+	});
+
+	it.only('Returns proper response for duplicate transaction', async () => {
```
Comment (sameersubudhi) [inline tests/integration/api_v3/rpc/dryRunTransactions.test.js:59]:
```suggestion
	it('Returns proper response for duplicate transaction', async () => {
```

File: tests/integration/api_v3/http/dryRunTransactions.test.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+			{ transaction: VALID_TRANSACTION },
+		);
+
+		expect(firstResponse).toMap(goodRequestSchema);
+		expect(firstResponse.data).toMap(dryrunTransactionResponseSchema);
+		expect(firstResponse.data.events.length).toBeGreaterThan(0);
+		expect(firstResponse.data.success).toBe(true);
+		expect(firstResponse.meta).toMap(metaSchema);
+
+		// Send transaction and wait to be minted
```
Comment (sameersubudhi) [inline tests/integration/api_v3/http/dryRunTransactions.test.js:68]:
```suggestion
		// Send transaction and wait for it to be included in the next block
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: tests/integration/api_v3/http/dryRunTransactions.test.js (hunk 15 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/http/dryRunTransactions.test.js:68]: applied.

File: tests/integration/api_v3/rpc/dryRunTransactions.test.js (hunk 14 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/rpc/dryRunTransactions.test.js:59]: applied.

File: tests/integration/api_v3/rpc/dryRunTransactions.test.js (hunk 13 above)
Comment (priojeetpriyom) [reply tests/integration/api_v3/rpc/dryRunTransactions.test.js:72]: applied.

File: tests/schemas/api_v3/dryrunTransaction.schema.js (hunk 11 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/dryrunTransaction.schema.js:27]: applied.

File: tests/schemas/api_v3/event.schema.js (hunk 10 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/event.schema.js:33]: applied.

File: tests/schemas/api_v3/regex.js (hunk 8 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/regex.js:35]: Added for events. topics are sent in response of dryrun.

File: tests/schemas/api_v3/dryrunTransaction.schema.js (hunk 12 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/dryrunTransaction.schema.js:31]: skipped as it makes the duplicate transaction test fail

File: tests/schemas/api_v3/event.schema.js (hunk 9 above)
Comment (priojeetpriyom) [reply tests/schemas/api_v3/event.schema.js]: applied.

File: tests/schemas/api_v3/dryrunTransaction.schema.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+const regex = require('./regex');
+
+const event = {
+	data: Joi.string().object().required(),
+	index: Joi.number().integer().min(0).required(),
+	module: Joi.string().pattern(regex.MODULE).required(),
+	name: Joi.string().pattern(regex.NAME).required(),
+	topics: Joi.array().items(Joi.string().pattern(regex.TOPIC)).required(),
+	height: Joi.number().positive().required(),
```
Comment (sameersubudhi) [inline tests/schemas/api_v3/dryrunTransaction.schema.js:26]: Please re-verify if the dry run events have the `height` property set? I believe not.

**Review by sameersubudhi — APPROVED** [review]
LGTM 👍

File: tests/utils.js [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const waitMs = (n) => new Promise((resolve) => {
+	setTimeout(() => {
+		resolve();
+	}, n);
+});
```
Comment (nagdahimanshu) [inline tests/utils.js:20]: Can we move the method to `tests/helpers/utils.js` as we already have `utils.js` file for tests
```suggestion
const waitMs = (n) => new Promise((resolve) => {
	setTimeout(() => {
		resolve();
	}, n);
});
```

Comment (priojeetpriyom) [reply tests/utils.js]: applied.

**Review by nagdahimanshu — APPROVED** [review]
Looks good 👍

**Pull request merged**